test, assumeutxo: Remove resolved todo comments and add new test

fjahr opened the pull request:
The first commit removes three Todos that have been addressed previously (see commit message for details).

The second message resolves another todo by adding the missing test case. This is a special case of "the tip has more work than the snapshot" where the tip is the same block as the snapshot base block.

Related to #28648.

fjahr committed:
test: Remove already resolved assumeutxo todo comments

- "Valid snapshot file, but referencing a snapshot block that turns out
  to be invalid, or has an invalid parent" has been addressed in #30267
- "An ancestor of snapshot block" - If chain tip refers to blocks in this context then any successful load is addressing this because if we have synced past the snapshot base block we fail because we don't need assumeutxo anymore. And if this is about headers then this is the `test_headers_not_synced()` case.
- "A descendant of the snapshot block" - If this refers to blocks the
  `test_snapshot_with_less_work()` addressed this and if it is just headers in this case again it would be represented in all of the successful loads in the test.

Co-authored-by: Alfonso Roman Zubeldia <alfonsoromanz24@gmail.com>

alfonsoromanz commented:
ACK 9d5a89509031ce924b9685a6e2e4b19cfef50a9e

I wonder if this is a good place to discuss other (potentially addressed) TODOs?

For example, interesting starting states could be loading a snapshot when the current chain tip is:

`- TODO: An ancestor of snapshot block`: isn't this already addressed when successfully loading the snapshot into node1 [here](https://github.com/bitcoin/bitcoin/blob/master/test/functional/feature_assumeutxo.py#L310)? Node1's chain tip is at `START_HEIGTH` (199) before loading the snapshot, and the block 199 is an ancestor of the snapshot block (299)

`- TODO: A descendant of the snapshot block`: isn't this addressed by the function `test_snapshot_with_less_work`? ([link](https://github.com/bitcoin/bitcoin/blob/master/test/functional/feature_assumeutxo.py#L201-L206)). Node0 is at `FINAL_HEIGHT` (399) with the tip being a descendant of the snapshot block (at `299`)

tdb3 reviewed (approved):
ACK 9d5a89509031ce924b9685a6e2e4b19cfef50a9e
Double checking https://github.com/bitcoin/bitcoin/pull/30403#issuecomment-2211817861 would also be good as well.

fjahr committed:
test: Add loadtxoutset test with tip on snapshot block

Also pulls out the guarding assert and calls it explicitly before the test function is called. This is already done before the existing call of the test function so it was not needed there.

fjahr commented:
> I wonder if this is a good place to discuss other (potentially addressed) TODOs?

Right, good idea. I avoided this for a while because I thought I might miss something because I misinterpret them but since we had these detailed discussions in #29996 I am now pretty confident we can remove some that are addressed.

> For example, interesting starting states could be loading a snapshot when the current chain tip is:
> 
> `- TODO: An ancestor of snapshot block`: isn't this already addressed when successfully loading the snapshot into node1 [here](https://github.com/bitcoin/bitcoin/blob/master/test/functional/feature_assumeutxo.py#L310)? Node1's chain tip is at `START_HEIGTH` (199) before loading the snapshot, and the block 199 is an ancestor of the snapshot block (299)

I think this is addressed and I am also unsure if this was ever not addressed :) Elsewhere I realized that I am not sure if "chain tip" in this comment refers to sync of headers or of blocks. But either way: If this is about blocks then any successful load is addressing this because if we have synced past the snapshot base block we fail because we don't need assumeutxo anymore. And if this is about headers then this is the `test_headers_not_synced()` case.

> `- TODO: A descendant of the snapshot block`: isn't this addressed by the function `test_snapshot_with_less_work`? ([link](https://github.com/bitcoin/bitcoin/blob/master/test/functional/feature_assumeutxo.py#L201-L206)). Node0 is at `FINAL_HEIGHT` (399) with the tip being a descendant of the snapshot block (at `299`)

Yeah, I agree and if it is just headers in this case again it would be represented in all of the successful loads in the test.

fjahr commented:
I have amended the commit to remove the previously named comments + made @alfonsoromanz co-author.

In the second commit I now resolve one of the few remaining cases where the tip is on the same block as the snapshot. This is a special case of the "more work" case so this was quite simple to do.

I think this means with this and the remaining open test PRs from @alfonsoromanz and @mzumsande we could resolve this list completely :)

fjahr commented:
To give more detail where they are resolved: I think #30320 resolves two of the remaining three, it's just not part of the PR yet, see: https://github.com/bitcoin/bitcoin/pull/30320#discussion_r1665811272 and #29996 removes the third as already done in the commit here: https://github.com/bitcoin/bitcoin/pull/29996/commits/5b7f70ba2661a194a3c476b81e03785feddb4e1c.

The last one to get merged gets to remove the whole comment section 🥳

alfonsoromanz reviewed (approved):
Re ACK 29d19c414c9e138c99f8de84e398528db92ff07e. The test execution is successful. 

I agree that this PR, along with [#30320](https://github.com/bitcoin/bitcoin/pull/30320) and [#29996](https://github.com/bitcoin/bitcoin/pull/29996), would resolve the TODO list completely.

Thanks!

BrandonOdiwuor reviewed (approved):
Code Review ACK 29d19c414c9e138c99f8de84e398528db92ff07e

I will close PR https://github.com/bitcoin/bitcoin/pull/29681 in favor of this

fjahr commented:
Oh, sorry @BrandonOdiwuor , I missed https://github.com/bitcoin/bitcoin/pull/29681 😞

tdb3 reviewed (approved):
ACK 29d19c414c9e138c99f8de84e398528db92ff07e

Left some nits.

fjahr commented on test/functional/feature_assumeutxo.py:
```diff
@@ -13,17 +13,12 @@
 
 Interesting starting states could be loading a snapshot when the current chain tip is:
 
-- TODO: An ancestor of snapshot block
```
I think it's a bit too verbose for the log but I added a comment. And I think it will be confusing to explicitly say everywhere in the normal case that the tip is an ancestor of the snapshot. But it's documented in one place now at least.

tdb3 replied:
Yup, seems like its the most likely use case for assumeutxo, so I could see how that could be quite verbose.

fjahr commented on test/functional/feature_assumeutxo.py:
```diff
@@ -267,6 +261,12 @@ def run_test(self):
+        # the snapshot
+        assert_equal(n0.getblockcount(), SNAPSHOT_BASE_HEIGHT)
+        assert_equal(n0.getblockchaininfo()["blocks"], SNAPSHOT_BASE_HEIGHT)
+        self.test_snapshot_with_less_work(dump_output['path'])
```
I think it's fine to keep the function name as is. The comment (now a log) provides the necessary context and I think it's just a special edge case of the less work behavior.

fjahr commented on test/functional/feature_assumeutxo.py:
```diff
@@ -267,6 +261,12 @@ def run_test(self):
         dump_output = n0.dumptxoutset('utxos.dat')
 
+        # Test special case where the tip of the node is on the same block as
+        # the snapshot
```
Done

alfonsoromanz commented on test/functional/feature_assumeutxo.py:
```diff
@@ -307,6 +306,8 @@ def run_test(self):
 
         self.log.info(f"Loading snapshot into second node from {dump_output['path']}")
+        # This node's tip is on an ancestor block of the snapshot, which should
+        # the normal case
```
**Nit:** Missing a "be", i.e. "which should *be* the normal case." Not sure if it's worth fixing, but just wanted to point it out

fjahr replied:
fixed, thanks!

alfonsoromanz reviewed (approved):
Re ACK e523bcfc3403700debe1f286fa8c3db2d347e40b

fjahr commented:
Rebased

tdb3 reviewed (approved):
ACK e1018672672f39910655ab37080bf3213ca55a39

alfonsoromanz reviewed (approved):
Re ACK e1018672672f39910655ab37080bf3213ca55a39

fjahr commented:
Rebased and since this is the last of the PRs addressing the TODOs comment in `feature_assumeutxo.py`, it now removes that whole section.

jrakibi commented:
ACK [d63ef73](https://github.com/bitcoin/bitcoin/commit/d63ef738001fb69ce04134cc8645dcd1e1cbccd1)

maflcko reviewed (approved):
ACK d63ef738001fb69ce04134cc8645dcd1e1cbccd1

left a nit (possibly a new test idea, but this can be done later).

alfonsoromanz reviewed (approved):
Re ACK d63ef738001fb69ce04134cc8645dcd1e1cbccd1

achow101 commented:
ACK d63ef738001fb69ce04134cc8645dcd1e1cbccd1
